src/main/java/com/fasterxml/jackson/databind/deser/std/DateDeserializers.java

```diff
@@ -69,6 +70,8 @@ protected abstract static class DateBasedDeserializer<T>
          */
         protected final DateFormat _customFormat;
 
+        private final ReentrantLock _customFormatLock = new ReentrantLock();
```

Comment:
This code is not 100% correct - If you have 2 deserializer instances with the same _customFormat instance then synchronizing it will prevent concurrent use of that _customFormat instance - your lock would not work properly in that case.

It is sort of annoying that Java never made DateFormat thread-safe. We could look into finding thread-safe alternatives - or adding a thread-safe class of our own that wraps DataFormat but that uses a locking mechanism to control access.

Reply:
I haven't researched this but I suspect we could find some `java.time` API to do date/time formatting that is thread safe - and therefore not need locking here at all.

Reply:
I agree, the existing `DateFormat` should be replaced with a `java.time.format.DateTimeFormatter` which is immutable and thread-safe and at that point this synchronization could be of course totally removed. However I guess that this could be addressed with a different pull request, immediately after this will be merged.

Reply:
I am not sure replacement would work, unfortunately, since we are talking about support for `java.util.Date` / `java.util.Calendar`; and if I remember correctly, format Strings used with `SimpleDateFormat` are different from those used with Java 8 date/time (and Joda).

So the better fix for users is to upgrade from java.util types to java.time.

So while I agree with all the ugliness of java.util world, I am not sure spending time on better support there is worth the effort.

Reply:
this Date code is not on critical path - users can avoid it using java.time classes instead of java.util Data/Calendar

so this could be left alone - and maybe tackled later